warehouse/query-core/src/main/java/datawave/query/jexl/visitors/PushdownNegationVisitor.java

```diff
@@ -0,0 +1,256 @@
+                // replace this node with an EQ node
+                JexlNode eqNode = JexlNodeFactory.buildEQNode(JexlASTHelper.getIdentifier(node), JexlASTHelper.getLiteral(node).image);
+                
+                JexlNodes.swap(node.jjtGetParent(), node, eqNode);
+                state.setNegated(false);
+                return state;
+            }
+        }
+        
+        // keep going
+        return super.visit(node, data);
+    }
+    
+    @Override
+    public Object visit(ASTNRNode node, Object data) {
+        if (data instanceof NegationState) {
+            NegationState state = (NegationState) data;
+            
+            if (state.isNegated()) {
+                // replace this node with an RE node
+                JexlNode eqNode = JexlNodeFactory.buildERNode(JexlASTHelper.getIdentifier(node), JexlASTHelper.getLiteral(node).image);
+                
+                JexlNodes.swap(node.jjtGetParent(), node, eqNode);
+                state.setNegated(false);
+                return state;
+            }
+        }
+        
+        // keep going
+        return super.visit(node, data);
+    }
+    
+    /**
+     * Apply De Morgan's law to a node, splicing out the old root and putting a replacement in its place. De Morgans law states that the negation of a
+     * disjunction is the conjunction of the negations; and the ne
+     *
+     * By generalizing De Morgans duality, an ASTAndNode can be converted to an ASTOrNode by applying a negation, flipping the operator (AND to OR vice versa),
+     * and negating each child then splicing the newly ro
+     * 
+     * @param root
+     *            the node to apply De Morgan's law to either an ASTAndNode or ASTOrNode, must not be null
+     * @return the replacement for root which was replaced, will always be an ASTNotNode
+     */
+    public static JexlNode applyDeMorgans(JexlNode root, boolean negateRoot) {
+        if (root == null) {
+            throw new IllegalArgumentException("root must not be null");
+        }
+        
+        if (root.jjtGetNumChildren() == 0) {
+            throw new IllegalStateException("root must have at least one child");
+        }
+        
+        // build the child list, while negating each child
+        List<JexlNode> negatedChildren = new ArrayList<>(root.jjtGetNumChildren());
+        for (int i = 0; i < root.jjtGetNumChildren(); i++) {
+            negatedChildren.add(JexlNodes.negate(root.jjtGetChild(i)));
+        }
+        
+        // flip the root operator and finish validating the root
+        JexlNode newRoot;
+        if (root instanceof ASTAndNode) {
+            newRoot = JexlNodeFactory.createOrNode(negatedChildren);
+        } else if (root instanceof ASTOrNode) {
+            newRoot = JexlNodeFactory.createAndNode(negatedChildren);
+        } else {
+            throw new IllegalArgumentException("root must be either an ASTAndNode or ASTOrNode");
+        }
+        
+        if (negateRoot) {
```

Comment:
Why would we ever want to negate the root?  It looks like you only ever pass 'true' for this param in the tests.

Reply:
Earlier implementations for this code did the full De Morgan's application. In the final version I left the negation of the root as an optional responsibility of the caller since the visitor doesn't actually need that, it just needs to be tracked for future recursive calls.